### DISCUSSION TITLE
Cannot establish communication from custom client

### DISCUSSION THREAD

**Discussion by srgjc** [microsoft/debugpy#1741] [Q&A] [answered]
I would like to communicate with debugpy from a custom client I am building over tcp locally.
For this I run `python -m debugpy --listen 5678 ./script.py`. (with or without `--wait-for-client` does not make a difference)
When listening to this port I get this response: 
```
Content-Length: 137

{"seq": 1, "type": "event", "event": "output", "body": {"category": "telemetry", "output": "ptvsd", "data": {"packageVersion": "1.8.8"}}}Content-Length: 139

{"seq": 2, "type": "event", "event": "output", "body": {"category": "telemetry", "output": "debugpy", "data": {"packageVersion": "1.8.8"}}}Content-Length: 189

{"seq": 3, "type": "event", "event": "debugpySockets", "body": {"sockets": [{"host": "127.0.0.1", "port": 5679, "internal": false}, {"host": "127.0.0.1", "port": 60323, "internal": true}]}}
```
Then I establish a connection to the port marked as internal and get the following request:
```
Content-Length: 168

{"seq": 1, "type": "request", "command": "pydevdAuthorize", "arguments": {"debugServerAccessToken": "76cec100048a7be9dc4f3195cefcf8d8687e56cde72ae065587cfc6ac659f1b4"}}
```
After looking at the source code I found how the json response should look like and sent this: (As for the token, I saw in the source it should match the server token, this might be wrong though...)
```
{"seq":"1","type":"response","request_seq":"1","success":true,"command":"pydevdAuthorize","body":{"clientAccessToken":"76cec100048a7be9dc4f3195cefcf8d8687e56cde72ae065587cfc6ac659f1b4"}}
```
After this I do not get a response.
Sending requests to the external ports also results in no reaction.
What is the correct way of establishing communication with debugpy?

**Comment by rchiodo** [comment]
Did you look at the order of events here (specifically the initialize part)?
https://microsoft.github.io/debug-adapter-protocol/overview

You should be sending an 'initialize' event.

**Comment by srgjc** [comment]
I think you mean I should send an 'initialize' request and not an event, right?. I have tried this already as specified in `debugProtocol.json`:
```
{"seq":"1","type":"request","command":"initialize","arguments":{"adapterID":"b9e965ddf2460b8ecdbe31eaf30d396f79e84b48a9d6ce66292132fb054ce016"}}
```
I get no response from neither internal or external ports.
As adapterID - which is the only mandatory argument - I passed the previously mentioned debugServerAccessToken, but this might be wrong... where do I otherwise get the adapterID?

Does one not get a response with an error message if the request is not as expected? Any response information would be useful. Sadly I don`t receive any response to any kind of request I try.

**Reply by rchiodo** [reply, marked as answer]
I guess I would turn logging on in VS code and see what messages it sends:
https://github.com/microsoft/debugpy/wiki/Enable-debugger-logs

I'm not sure where it sends the access token. Enabling those debugger logs can also be done in your scenario, then debugpy (and pydevd) should output more information.

**Reply by srgjc** [reply]
I am not using vs code. As I mentioned at the beginning, I am trying to communicate with debugpy from a custom client that I am building. A java debug client with a gui. Or is this not possible?

**Reply by rchiodo** [reply]
Yes but I'm assuming you can use VS code to debug the same thing? You can then use that as an example to see what messages are being sent.

**Reply by rchiodo** [reply]
Additionally, if you enable the environment variables for logging in your scenario (where you put the listen and try to send your own messages), debugpy should output more information than just the messages you get as responses. You can also add extra logging to debugpy to see why it doesn't answer your initialize request.

**Reply by srgjc** [reply]
Thanks for clarifying, I get what you mean now. I will try your suggestions out and post the results here. Thank you for your time!

**Comment by srgjc** [comment]
Hello @rchiodo, I just wanted to thank you. I have come back to my project after months and I was able to resolve my issue. I had misconceptions about which socket was meant for the client and which for the debugpy server, and I was not aware of the debugpy adapter-server-pydevd dynamic. Using VS Code to attach to debugpy and watching the logs helped me mimic the same communication back and forth from my custom java client prototype and was able to initialize, attach and set a breakpoint. Thank you very much for the support!
